AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ValidatorParameterHelper.java

```diff
@@ -211,96 +255,115 @@ public void validate(ExecutableElement executableElement, IsValid valid) {
 		}
 
 		@Override
-		public ChainableParameterRequirement<AnyOrderParameterValidator> extendsType(String fullyQualifiedName) {
-			return new ChainableParameterRequirement<AnyOrderParameterValidator>(this, addExpectedParameter(fullyQualifiedName, true, true));
-		}
-
-		@Override
-		public ChainableParameterRequirement<AnyOrderParameterValidator> type(String fullyQualifiedName) {
-			return new ChainableParameterRequirement<AnyOrderParameterValidator>(this, addExpectedParameter(fullyQualifiedName, true, false));
-		}
-
-		@Override
 		protected String createMessage(ExecutableElement element) {
 			return super.createMessage(element) + " in any order";
 		}
 	}
 
-	public class InOrderParameterValidator extends ParameterValidatorBase<ChainableParameterRequirement<InOrderParameterValidator>> {
 
-		@Override
-		public ChainableParameterRequirement<InOrderParameterValidator> extendsType(String fullyQualifiedName) {
-			return new ChainableParameterRequirement<InOrderParameterValidator>(this, addExpectedParameter(fullyQualifiedName, true, true));
-		}
+	public class ParameterRequirement {
 
-		@Override
-		public ChainableParameterRequirement<InOrderParameterValidator> type(String fullyQualifiedName) {
-			return new ChainableParameterRequirement<InOrderParameterValidator>(this, addExpectedParameter(fullyQualifiedName, true, false));
-		}
+		private String typeName;
+		private boolean extending;
+		private Class<? extends Annotation> annotationClass;
+		private boolean required = true;
+		private boolean multiple = false;
+		private boolean hasBeenSatisfied = false;
 
-		@Override
-		protected String createMessage(ExecutableElement element) {
-			return super.createMessage(element) + " in the order above";
+		public ParameterRequirement(String typeName, boolean extending) {
+			this.typeName = typeName;
+			this.extending = extending;
 		}
-	}
 
-	public class OneParamValidator extends ParameterValidatorBase<ParameterRequirement<Validator>> {
-
-		@Override
-		public ParameterRequirement<Validator> extendsType(String fullyQualifiedName) {
-			return new ParameterRequirement<Validator>(this, addExpectedParameter(fullyQualifiedName, true, true));
+		public ParameterRequirement(Class<? extends Annotation> annotationClass) {
+			this.annotationClass = annotationClass;
 		}
 
-		@Override
-		public ParameterRequirement<Validator> type(String fullyQualifiedName) {
-			return new ParameterRequirement<Validator>(this, addExpectedParameter(fullyQualifiedName, true, false));
+		public void multiple() {
+			multiple = true;
 		}
 
-	}
+		public void optional() {
+			required = false;
+		}
 
-	public class NoParamValidator implements Validator {
+		public boolean isSatisfied(VariableElement param) {
```

Comment:
I am not sure. Maybe checked and satisfied? It is not much better. We
should use different verbs.